File: verifier-core/src/main/java/com/jetbrains/pluginverifier/verifiers/instruction/MultiANewArrayInstructionVerifier.java
package com.jetbrains.pluginverifier.verifiers.instruction;

import com.intellij.structure.resolvers.Resolver;
import com.jetbrains.pluginverifier.api.VContext;
import com.jetbrains.pluginverifier.location.ProblemLocation;
import com.jetbrains.pluginverifier.problems.ClassNotFoundProblem;
import com.jetbrains.pluginverifier.utils.VerifierUtil;
import org.objectweb.asm.tree.AbstractInsnNode;
import org.objectweb.asm.tree.ClassNode;
import org.objectweb.asm.tree.MethodNode;
import org.objectweb.asm.tree.MultiANewArrayInsnNode;

/**
 * @author Sergey Patrikeev
 */
public class MultiANewArrayInstructionVerifier implements InstructionVerifier {
  @Override
  public void verify(ClassNode clazz, MethodNode method, AbstractInsnNode instr, Resolver resolver, VContext ctx) {
    if (!(instr instanceof MultiANewArrayInsnNode)) return;
    MultiANewArrayInsnNode newArrayInstruction = (MultiANewArrayInsnNode) instr;
    String descr = VerifierUtil.extractClassNameFromDescr(newArrayInstruction.desc);
    if (descr == null) return;
    if (!VerifierUtil.classExistsOrExternal(ctx, resolver, descr)) {
      ctx.registerProblem(new ClassNotFoundProblem(descr), ProblemLocation.fromMethod(clazz.name, method));
    }
  }
}

File: verifier-core/src/main/java/com/jetbrains/pluginverifier/verifiers/instruction/MultiANewArrayInstructionVerifier.kt
package com.jetbrains.pluginverifier.verifiers.instruction

import com.jetbrains.pluginverifier.verifiers.VerificationContext
import com.jetbrains.pluginverifier.verifiers.createMethodLocation
import com.jetbrains.pluginverifier.verifiers.extractClassNameFromDescr
import com.jetbrains.pluginverifier.verifiers.resolveClassOrProblem
import org.objectweb.asm.tree.AbstractInsnNode
import org.objectweb.asm.tree.ClassNode
import org.objectweb.asm.tree.MethodNode
import org.objectweb.asm.tree.MultiANewArrayInsnNode

class MultiANewArrayInstructionVerifier : InstructionVerifier {
  override fun verify(clazz: ClassNode, method: MethodNode, instr: AbstractInsnNode, ctx: VerificationContext) {
    if (instr !is MultiANewArrayInsnNode) return
    val descr = instr.desc.extractClassNameFromDescr() ?: return

    //During resolution of the symbolic reference to the class, array, or interface type,
    // any of the exceptions documented in §5.4.3.1 can be thrown.
    ctx.resolveClassOrProblem(descr, clazz) { createMethodLocation(clazz, method) }
  }
}
